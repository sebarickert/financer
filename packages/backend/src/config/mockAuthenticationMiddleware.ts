import { Role, User } from '@prisma/client';
import { NextFunction, Request, Response } from 'express';

export const DUMMY_TEST_USER: Partial<User> = {
  roles: [Role.TEST_USER],
  id: '61460d7354ea082ad0256749',
  name: 'Dummy Test User',
};

export const mockAuthenticationMiddleware = (
  req: Request,
  _: Response,
  next: NextFunction,
): void => {
  req.user = DUMMY_TEST_USER;
  req.isAuthenticated = () => true;
  next();
};
