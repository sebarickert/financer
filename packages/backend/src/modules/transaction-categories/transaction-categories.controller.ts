import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
  Query,
} from '@nestjs/common';
import { TransactionType } from '@prisma/client';
import {
  ApiBody,
  ApiExtraModels,
  ApiOkResponse,
  ApiParam,
  ApiQuery,
  ApiTags,
} from '@silte/nestjs-swagger';

import { ValidateEntityId } from '../../utils/validate-entity-id.pipe';
import { LoggedIn } from '../auth/decorators/loggedIn.decorators';
import { UserId } from '../users/users.decorators';

import { CreateTransactionCategoryDto } from './dto/create-transaction-category.dto';
import { TransactionCategoryDto } from './dto/transaction-category.dto';
import { CategoryMonthlySummaryDto } from './dto/transaction-month-summary.dto';
import { UpdateTransactionCategoryDto } from './dto/update-transaction-category.dto';
import { TransactionCategoriesService } from './transaction-categories.service';

@Controller('api/transaction-categories')
@ApiTags('Transaction categories')
@ApiExtraModels(TransactionCategoryDto, CategoryMonthlySummaryDto)
@LoggedIn()
export class TransactionCategoriesController {
  constructor(
    private readonly transactionCategoriesService: TransactionCategoriesService,
  ) {}

  @Post()
  @ApiBody({ type: CreateTransactionCategoryDto })
  @ApiOkResponse({ schema: { properties: { payload: { type: 'string' } } } })
  async create(
    @UserId() userId: string,
    @Body() createTransactionCategoryDto: CreateTransactionCategoryDto,
  ) {
    return this.transactionCategoriesService.create(
      userId,
      createTransactionCategoryDto,
    );
  }

  @Get()
  @ApiOkResponse({
    type: [TransactionCategoryDto],
    description: 'Return transaction category by id',
  })
  @ApiQuery({
    name: 'visibilityType',
    required: false,
    enum: TransactionType,
    enumName: 'visibilityType',
  })
  findAllByUser(
    @UserId() userId: string,
    @Query('visibilityType') visibilityType?: TransactionType,
  ) {
    return this.transactionCategoriesService.findAllByUser(
      userId,
      visibilityType,
    );
  }

  @Get(':id')
  @ApiOkResponse({
    type: TransactionCategoryDto,
  })
  @ApiParam({
    name: 'id',
    type: String,
  })
  findOne(@UserId() userId: string, @Param('id', ValidateEntityId) id: string) {
    return this.transactionCategoriesService.findOne(userId, id);
  }

  @Get(':id/summary')
  @ApiOkResponse({
    type: [CategoryMonthlySummaryDto],
  })
  @ApiParam({
    name: 'id',
    type: String,
  })
  getCategorySummary(
    @UserId() userId: string,
    @Param('id', ValidateEntityId) id: string,
  ) {
    return this.transactionCategoriesService.findMonthlySummariesByUserAndId(
      userId,
      id,
    );
  }

  @Patch(':id')
  @ApiBody({ type: UpdateTransactionCategoryDto })
  @ApiOkResponse({ type: TransactionCategoryDto })
  @ApiParam({
    name: 'id',
    type: String,
  })
  update(
    @UserId() userId: string,
    @Param('id', ValidateEntityId) id: string,
    @Body() updateTransactionCategoryDto: UpdateTransactionCategoryDto,
  ) {
    return this.transactionCategoriesService.update(
      userId,
      id,
      updateTransactionCategoryDto,
    );
  }

  @Delete(':id')
  @ApiParam({
    name: 'id',
    type: String,
  })
  remove(@UserId() userId: string, @Param('id', ValidateEntityId) id: string) {
    return this.transactionCategoriesService.remove(userId, id);
  }
}
