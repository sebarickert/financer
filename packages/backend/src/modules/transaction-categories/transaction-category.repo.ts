import { Injectable } from '@nestjs/common';
import { Prisma, PrismaPromise, TransactionCategory } from '@prisma/client';

import { PrismaService } from '@/database/prisma.service';

@Injectable()
export class TransactionCategoryRepo {
  constructor(private readonly prisma: PrismaService) {}

  async findOne(
    transactionCategoryWhereUniqueInput: Prisma.TransactionCategoryWhereUniqueInput,
  ): Promise<TransactionCategory | null> {
    return this.prisma.transactionCategory.findUnique({
      where: transactionCategoryWhereUniqueInput,
    });
  }

  async findMany(params: {
    skip?: number;
    take?: number;
    cursor?: Prisma.TransactionCategoryWhereUniqueInput;
    where?: Prisma.TransactionCategoryWhereInput;
    orderBy?: Prisma.TransactionCategoryOrderByWithRelationInput;
  }): Promise<TransactionCategory[]> {
    const { skip, take, cursor, where, orderBy } = params;
    return this.prisma.transactionCategory.findMany({
      skip,
      take,
      cursor,
      where,
      orderBy,
    });
  }

  async getCount(params: {
    skip?: number;
    take?: number;
    cursor?: Prisma.TransactionCategoryWhereUniqueInput;
    where?: Prisma.TransactionCategoryWhereInput;
    orderBy?: Prisma.TransactionCategoryOrderByWithRelationInput;
  }): Promise<number> {
    const { skip, take, cursor, where, orderBy } = params;
    return this.prisma.transactionCategory.count({
      skip,
      take,
      cursor,
      where,
      orderBy,
    });
  }

  async create(
    data: Prisma.TransactionCategoryUncheckedCreateInput,
  ): Promise<TransactionCategory> {
    return this.prisma.transactionCategory.create({
      data: {
        id: data.id,
        name: data.name,
        userId: data.userId,
        parentCategoryId: data.parentCategoryId,
        visibility: data.visibility,
        deleted: data.deleted,
      },
    });
  }

  createMany(
    data: Prisma.TransactionCategoryUncheckedCreateInput[],
  ): PrismaPromise<Prisma.BatchPayload> {
    return this.prisma.transactionCategory.createMany({
      data: data.map((item) => ({
        id: item.id,
        name: item.name,
        userId: item.userId,
        parentCategoryId: item.parentCategoryId,
        visibility: item.visibility,
        deleted: item.deleted,
      })),
    });
  }

  async update(params: {
    where: Prisma.TransactionCategoryWhereUniqueInput;
    data: Prisma.TransactionCategoryUpdateInput;
  }): Promise<TransactionCategory> {
    const { where, data } = params;
    return this.prisma.transactionCategory.update({
      data: {
        id: data.id,
        name: data.name,
        userId: data.userId,
        parentCategoryId: data.parentCategoryId,
        visibility: data.visibility,
        deleted: data.deleted,
      },
      where,
    });
  }

  async delete(
    where: Prisma.TransactionCategoryWhereUniqueInput,
  ): Promise<TransactionCategory> {
    return this.prisma.transactionCategory.delete({
      where,
    });
  }

  deleteMany(where: Prisma.TransactionCategoryWhereInput) {
    return this.prisma.transactionCategory.deleteMany({
      where,
    });
  }
}
