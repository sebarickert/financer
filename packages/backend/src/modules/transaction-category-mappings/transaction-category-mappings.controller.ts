import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
} from '@nestjs/common';

import { ObjectId } from '../../types/objectId';
import { LoggedIn } from '../auth/decorators/loggedIn.decorators';
import { UserId } from '../users/users.decorators';

import { CreateTransactionCategoryMappingDto } from './dto/create-transaction-category-mapping.dto';
import { UpdateTransactionCategoryMappingDto } from './dto/update-transaction-category-mapping.dto';
import { TransactionCategoryMappingsService } from './transaction-category-mappings.service';

@Controller('api/transaction-category-mappings')
@LoggedIn()
export class TransactionCategoryMappingsController {
  constructor(
    private readonly transactionCategoryMappingsService: TransactionCategoryMappingsService,
  ) {}

  @Post()
  create(
    @Body()
    createTransactionCategoryMappingDto: CreateTransactionCategoryMappingDto,
  ) {
    return this.transactionCategoryMappingsService.create(
      createTransactionCategoryMappingDto,
    );
  }

  @Get()
  findAllByUser(@UserId() userId: ObjectId) {
    return this.transactionCategoryMappingsService.findAllByUser(userId);
  }

  @Get(':id')
  findOne(@Param('id') id: string) {
    return this.transactionCategoryMappingsService.findOne(+id);
  }

  @Patch(':id')
  update(
    @Param('id') id: string,
    @Body()
    updateTransactionCategoryMappingDto: UpdateTransactionCategoryMappingDto,
  ) {
    return this.transactionCategoryMappingsService.update(
      +id,
      updateTransactionCategoryMappingDto,
    );
  }

  @Delete(':id')
  remove(@Param('id') id: string) {
    return this.transactionCategoryMappingsService.remove(+id);
  }
}
