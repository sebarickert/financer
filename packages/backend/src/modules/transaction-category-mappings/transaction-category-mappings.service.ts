import { BadRequestException, Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { TransactionCategoryMapping, TransactionType } from '@prisma/client';
import { Model } from 'mongoose';

import { TransactionCategoryMappingRepo } from '../../database/repos/transaction-category-mapping.repo';
import { ObjectId, parseObjectId } from '../../types/objectId';
import { CategoryMonthlySummaryDto } from '../transaction-categories/dto/transaction-month-summary.dto';

import { CreateTransactionCategoryMappingDto } from './dto/create-transaction-category-mapping.dto';
import {
  TransactionCategoryMapping as TransactionCategoryMappingOld,
  TransactionCategoryMappingDocument,
} from './schemas/transaction-category-mapping.schema';

@Injectable()
export class TransactionCategoryMappingsService {
  constructor(
    private readonly transactionCategoryMappingRepo: TransactionCategoryMappingRepo,
    @InjectModel(TransactionCategoryMappingOld.name)
    private transactionCategoryMappingModel: Model<TransactionCategoryMappingDocument>,
  ) {}

  async createMany(
    userId: string,
    createTransactionCategoryMappingDto: CreateTransactionCategoryMappingDto[],
  ) {
    return this.transactionCategoryMappingRepo.createMany(
      createTransactionCategoryMappingDto.map((category) => ({
        ...category,
        userId,
      })),
    );
  }

  findAll() {
    return `This action returns all transactionCategoryMappings`;
  }

  findOne(id: number) {
    return `This action returns a #${id} transactionCategoryMapping`;
  }

  async findAllByUser(userId: string): Promise<TransactionCategoryMapping[]> {
    return this.transactionCategoryMappingRepo.findMany({ where: { userId } });
  }

  async findAllByUserAndCategoryIds(
    userId: string,
    categoryIds: string[],
  ): Promise<TransactionCategoryMapping[]> {
    return this.transactionCategoryMappingRepo.findMany({
      where: {
        userId,
        categoryId: {
          in: categoryIds,
        },
      },
    });
  }

  async findAllByUserAndTransaction(
    userId: string,
    transactionId: string,
  ): Promise<TransactionCategoryMapping[]> {
    return this.transactionCategoryMappingRepo.findMany({
      where: { userId, transactionId },
    });
  }

  async findMonthlySummariesByUserAndId(
    userIdStr: string,
    categoryIds: string[],
    limit?: number,
    year?: number,
    month?: number,
  ): Promise<CategoryMonthlySummaryDto[]> {
    const userId = parseObjectId(userIdStr);

    return this.transactionCategoryMappingModel
      .aggregate([
        {
          $match: {
            owner: userId,
            category_id: { $in: categoryIds },
            ...this.getYearAndMonthFilter(year, month, 'laterThan'),
          },
        },
        {
          $lookup: {
            from: 'transactions',
            localField: 'transaction_id',
            foreignField: '_id',
            as: 'transaction',
          },
        },
        {
          $project: {
            date: {
              $arrayElemAt: ['$transaction.date', 0],
            },
            amount: '$amount',
            transactionAmount: {
              $arrayElemAt: ['$transaction.amount', 0],
            },
            fromAccount: {
              $arrayElemAt: ['$transaction.fromAccount', 0],
            },
            toAccount: {
              $arrayElemAt: ['$transaction.toAccount', 0],
            },
          },
        },
        {
          $group: {
            _id: {
              year: { $year: '$date' },
              month: { $month: '$date' },
            },
            totalCount: {
              $sum: await this.getMonthlySummaryCondition(userId, 1, null),
            },
            incomeCount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                1,
                TransactionType.INCOME,
              ),
            },
            expenseCount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                1,
                TransactionType.EXPENSE,
              ),
            },
            transferCount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                1,
                TransactionType.TRANSFER,
              ),
            },
            totalAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$amount',
                null,
              ),
            },
            totalTransactionAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$transactionAmount',
                null,
              ),
            },
            incomeAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$amount',
                TransactionType.INCOME,
              ),
            },
            incomeTransactionAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$transactionAmount',
                TransactionType.INCOME,
              ),
            },
            expenseAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$amount',
                TransactionType.EXPENSE,
              ),
            },
            expenseTransactionAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$transactionAmount',
                TransactionType.EXPENSE,
              ),
            },
            transferAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$amount',
                TransactionType.TRANSFER,
              ),
            },
            transferTransactionAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$transactionAmount',
                TransactionType.TRANSFER,
              ),
            },
          },
        },
        {
          $sort: {
            _id: -1,
          },
        },
        {
          $project: {
            'total.count': '$totalCount',
            'total.amount': '$totalAmount',
            'total.transactionAmount': '$totalTransactionAmount',
            'income.count': '$incomeCount',
            'income.amount': '$incomeAmount',
            'income.transactionAmount': '$incomeTransactionAmount',
            'expense.count': '$expenseCount',
            'expense.amount': '$expenseAmount',
            'expense.transactionAmount': '$expenseTransactionAmount',
            'transfer.count': '$transferCount',
            'transfer.amount': '$transferAmount',
            'transfer.transactionAmount': '$transferTransactionAmount',
          },
        },
      ])
      .limit(limit || 1000)
      .exec();
  }

  private getYearAndMonthFilter(
    year?: number,
    month?: number,
    filterMode: 'targetMonth' | 'laterThan' = 'targetMonth',
  ) {
    if (!year && month) {
      throw new BadRequestException('Year is required when month is provided');
    }

    if (!year && !month) {
      return {};
    }

    if (filterMode === 'laterThan') {
      return {
        date: {
          $gte: new Date(year, month - 1 || 0, 1),
        },
      };
    }

    return {
      date: {
        $gte: new Date(year, month - 1 || 0, 1),
        $lt: new Date(year, month || 12, 1),
      },
    };
  }

  private async getMonthlySummaryCondition(
    userId: ObjectId,
    operator: '$amount' | '$transactionAmount' | 1 | 0,
    transactionType: TransactionType,
  ) {
    const selectedQuery =
      this.getAggregationTransactionTypeFilter(transactionType);

    if (transactionType !== null || typeof operator !== 'string') {
      return { $cond: [{ $and: selectedQuery }, operator, 0] };
    }

    return {
      $cond: [
        { $and: [...selectedQuery, { $eq: ['$fromAccount', null] }] },
        operator,
        {
          $cond: [
            { $and: [...selectedQuery, { $eq: ['$toAccount', null] }] },
            { $multiply: [operator, -1] },
            0,
          ],
        },
      ],
    };
  }

  private getAggregationTransactionTypeFilter(
    transactionType: TransactionType,
  ): { [key in string]: unknown }[] {
    const isEmpty = (fieldName) => ({ $eq: [fieldName, undefined] });
    const isNotEmpty = (fieldName) => ({ $ne: [fieldName, undefined] });

    switch (transactionType) {
      case TransactionType.INCOME:
        return [isNotEmpty('$toAccount'), isEmpty('$fromAccount')];
      case TransactionType.EXPENSE:
        return [isNotEmpty('$fromAccount'), isEmpty('$toAccount')];
      case TransactionType.TRANSFER:
        return [isNotEmpty('$fromAccount'), isNotEmpty('$toAccount')];
      case null:
        return [];
      default:
        throw new Error(`Invalid transaction type: ${transactionType}`);
    }
  }

  update(id: number) {
    return `This action updates a #${id} transactionCategoryMapping`;
  }

  remove(id: number) {
    return `This action removes a #${id} transactionCategoryMapping`;
  }

  async removeAllByUserAndTransaction(userId: string, transactionId: string) {
    await this.transactionCategoryMappingRepo.deleteMany({
      userId,
      transactionId,
    });
  }

  async removeAllByUser(userId: string) {
    await this.transactionCategoryMappingRepo.deleteMany({
      userId,
    });
  }
}
