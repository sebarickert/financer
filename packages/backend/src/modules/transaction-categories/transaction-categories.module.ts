import { Module } from '@nestjs/common';

import { DatabaseModule } from '../../database/database.module';
import { TransactionCategoryMappingsModule } from '../transaction-category-mappings/transaction-category-mappings.module';

import { TransactionCategoriesController } from './transaction-categories.controller';
import { TransactionCategoriesService } from './transaction-categories.service';

@Module({
  imports: [DatabaseModule, TransactionCategoryMappingsModule],
  controllers: [TransactionCategoriesController],
  providers: [TransactionCategoriesService],
  exports: [TransactionCategoriesService],
})
export class TransactionCategoriesModule {}
