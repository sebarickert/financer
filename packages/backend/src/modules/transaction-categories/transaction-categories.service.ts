import {
  BadRequestException,
  Injectable,
  InternalServerErrorException,
  NotFoundException,
} from '@nestjs/common';
import { TransactionCategory, TransactionType } from '@prisma/client';

import { TransactionCategoryRepo } from '../../database/repos/transaction-category.repo';
import { TransactionCategoryMappingsService } from '../transaction-category-mappings/transaction-category-mappings.service';

import { CreateTransactionCategoryDto } from './dto/create-transaction-category.dto';
import { CategoryMonthlySummaryDto } from './dto/transaction-month-summary.dto';
import { UpdateTransactionCategoryDto } from './dto/update-transaction-category.dto';

@Injectable()
export class TransactionCategoriesService {
  constructor(
    private readonly transactionCategoryRepo: TransactionCategoryRepo,
    private readonly transactionCategoryMappingsService: TransactionCategoryMappingsService,
  ) {}

  private async isParentIdInChildHierarchy(
    categoryId: string,
    newParentId: string,
  ): Promise<boolean> {
    const categories = await this.findAllChildrensById([categoryId]);
    const ids = categories.map(({ id }) => id);

    return ids.includes(newParentId);
  }

  async create(
    userId: string,
    createTransactionCategoryDto: CreateTransactionCategoryDto,
  ): Promise<TransactionCategory> {
    const newCategory = {
      ...createTransactionCategoryDto,
      userId,
    };

    if (newCategory.parentCategoryId) {
      await this.findOne(userId, newCategory.parentCategoryId);
    }

    return this.transactionCategoryRepo.create(newCategory);
  }

  async createMany(
    userId: string,
    createTransactionCategoryDto: CreateTransactionCategoryDto[],
  ) {
    return this.transactionCategoryRepo.createMany(
      createTransactionCategoryDto.map((category) => ({ ...category, userId })),
    );
  }

  findAll() {
    return `This action returns all transactionCategories`;
  }

  async findOne(userId: string, id: string): Promise<TransactionCategory> {
    const category = await this.transactionCategoryRepo.findOne({ id, userId });

    if (!category) {
      throw new NotFoundException('Category not found.');
    }

    return category;
  }

  async findAllChildrensById(parentIds: string[], depth = 0) {
    if (depth > 10) {
      throw new InternalServerErrorException(
        'Too many levels of nesting, probably infite loop on hierarchy?',
      );
    }

    const categories = await this.transactionCategoryRepo.findMany({
      where: {
        parentCategoryId: { in: parentIds },
      },
    });

    const ids = categories.map((category) => category.id);

    if (ids.length === 0) {
      return categories;
    }

    const childCategories = await this.findAllChildrensById(ids, depth + 1);

    return [...categories, ...childCategories];
  }

  async findAllByUser(
    userId: string,
    visibilityType?: TransactionType,
  ): Promise<TransactionCategory[]> {
    return this.transactionCategoryRepo.findMany({
      where: {
        userId,
        deleted: { not: true },
        visibility: {
          has: visibilityType,
        },
      },
    });
  }

  async findMonthlySummariesByUserAndId(
    userId: string,
    parentCategoryId: string,
    year?: number,
    month?: number,
  ): Promise<CategoryMonthlySummaryDto[]> {
    const childrenIds = await this.findAllChildrensById([parentCategoryId]);
    const targetIds = [parentCategoryId, ...childrenIds];

    return this.transactionCategoryMappingsService.findMonthlySummariesByUserAndId(
      userId,
      targetIds,
      year,
      month,
    );
  }

  async update(
    userId: string,
    id: string,
    updateTransactionCategoryDto: UpdateTransactionCategoryDto,
  ) {
    await this.findOne(userId, id);

    if (updateTransactionCategoryDto.parentCategoryId) {
      const parentId = updateTransactionCategoryDto.parentCategoryId;

      if (await this.isParentIdInChildHierarchy(id, parentId)) {
        throw new BadRequestException(
          'Parent category cannot be child category of current item.',
        );
      }
    }

    return this.transactionCategoryRepo.update({
      where: { id, userId },
      data: updateTransactionCategoryDto,
    });
  }

  async remove(userId: string, id: string) {
    await this.findOne(userId, id);

    await this.transactionCategoryRepo.update({
      where: { id, userId },
      data: { deleted: true },
    });
  }

  async removeAllByUser(userId: string) {
    await this.transactionCategoryRepo.deleteMany({ userId });
  }

  async ensureCategoriesExist(ids: string[]) {
    const categories = await this.transactionCategoryRepo.findMany({
      where: {
        id: {
          in: ids,
        },
      },
    });

    const hasMatchingCategoryIds = ids.every((id) =>
      categories.some((category) => category.id === id),
    );

    if (!hasMatchingCategoryIds) {
      throw new BadRequestException('One or more categories does not exist.');
    }
  }
}
