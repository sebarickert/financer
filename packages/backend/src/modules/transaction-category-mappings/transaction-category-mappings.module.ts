import { Module } from '@nestjs/common';

import { DatabaseModule } from '../../database/database.module';

import { TransactionCategoryMappingsService } from './transaction-category-mappings.service';

@Module({
  imports: [DatabaseModule],
  providers: [TransactionCategoryMappingsService],
  exports: [TransactionCategoryMappingsService],
})
export class TransactionCategoryMappingsModule {}
