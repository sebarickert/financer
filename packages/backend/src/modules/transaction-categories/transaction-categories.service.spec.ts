import { Test, TestingModule } from '@nestjs/testing';

import { createMockServiceProvider } from '../../../test/create-mock-service-provider';
import { DUMMY_TEST_USER } from '../../config/mockAuthenticationMiddleware';
import { transactionCategoryMappingRepoFindAllByUserIdTransactionData } from '../../database/repos/mocks/transaction-category-mapping-repo-mock';
import {
  transactionCategoryRepoMockDataFindById,
  transactionCategoryRepoUserMockDataFindAllBy,
} from '../../database/repos/mocks/transaction-category-repo-mock';
import { TransactionCategoryMappingRepo } from '../../database/repos/transaction-category-mapping.repo';
import { TransactionCategoryRepo } from '../../database/repos/transaction-category.repo';
import { TransactionCategoryMappingsService } from '../transaction-category-mappings/transaction-category-mappings.service';

import { TransactionCategoriesService } from './transaction-categories.service';

describe('TransactionCategoriesService', () => {
  let service: TransactionCategoriesService;
  let transactionCategoryRepo: jest.Mocked<TransactionCategoryRepo>;
  let transactionCategoryMappingRepo: jest.Mocked<TransactionCategoryMappingRepo>;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        TransactionCategoriesService,
        TransactionCategoryMappingsService,
        createMockServiceProvider(TransactionCategoryRepo),
        createMockServiceProvider(TransactionCategoryMappingRepo),
      ],
    }).compile();

    service = module.get<TransactionCategoriesService>(
      TransactionCategoriesService,
    );
    transactionCategoryRepo = module.get<jest.Mocked<TransactionCategoryRepo>>(
      TransactionCategoryRepo,
    );
    transactionCategoryMappingRepo = module.get<
      jest.Mocked<TransactionCategoryMappingRepo>
    >(TransactionCategoryMappingRepo);
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should return a transaction category by id', async () => {
    const id = '623b58ada3deba9879422fbf';
    jest
      .spyOn(transactionCategoryRepo, 'findOne')
      .mockResolvedValue(transactionCategoryRepoMockDataFindById[id]);

    jest
      .spyOn(transactionCategoryRepo, 'findMany')
      .mockResolvedValue(transactionCategoryRepoUserMockDataFindAllBy);

    const transactionCategory = await service.findOne(DUMMY_TEST_USER.id, id);

    expect(transactionCategoryRepo.findOne).toHaveBeenCalledTimes(1);
    expect(transactionCategoryRepo.findOne).toHaveBeenCalledWith({
      id: '623b58ada3deba9879422fbf',
      userId: '61460d7354ea082ad0256749',
    });

    expect(transactionCategoryRepo.findMany).toHaveBeenCalledTimes(1);
    expect(transactionCategoryRepo.findMany).toHaveBeenCalledWith({
      where: {
        deleted: {
          not: true,
        },
        userId: '61460d7354ea082ad0256749',
        visibility: undefined,
      },
    });

    expect(transactionCategory).toMatchSnapshot();
  });

  it('should return an array of transaction categories from findAllByUser', async () => {
    jest
      .spyOn(transactionCategoryRepo, 'findMany')
      .mockResolvedValue(transactionCategoryRepoUserMockDataFindAllBy);

    const transactionCategories = await service.findAllByUser(
      DUMMY_TEST_USER.id,
    );

    expect(transactionCategoryRepo.findMany).toHaveBeenCalledTimes(1);
    expect(transactionCategoryRepo.findMany).toHaveBeenCalledWith({
      where: {
        deleted: {
          not: true,
        },
        userId: '61460d7354ea082ad0256749',
        visibility: undefined,
      },
    });

    expect(transactionCategories).toMatchSnapshot();
  });

  it('should return an array of transaction categories from findMonthlySummariesByUserAndId', async () => {
    jest.spyOn(transactionCategoryRepo, 'findMany').mockResolvedValue([]);
    jest
      .spyOn(transactionCategoryMappingRepo, 'findMany')
      .mockResolvedValue(
        transactionCategoryMappingRepoFindAllByUserIdTransactionData,
      );

    const transactionCategories = await service.findMonthlySummariesByUserAndId(
      DUMMY_TEST_USER.id,
      '623b58ada3deba9879422fbf',
    );

    expect(transactionCategoryRepo.findMany).toHaveBeenCalledTimes(1);
    expect(transactionCategoryRepo.findMany).toHaveBeenCalledWith({
      where: {
        parentCategoryId: {
          in: ['623b58ada3deba9879422fbf'],
        },
      },
    });

    expect(transactionCategoryMappingRepo.findMany).toHaveBeenCalledTimes(1);
    expect(transactionCategoryMappingRepo.findMany).toHaveBeenCalledWith({
      include: {
        transaction: {
          select: {
            date: true,
            fromAccount: true,
            toAccount: true,
          },
        },
      },
      where: {
        categoryId: {
          in: ['623b58ada3deba9879422fbf'],
        },
        userId: '61460d7354ea082ad0256749',
      },
    });

    expect(transactionCategories).toMatchSnapshot();
  });
});
