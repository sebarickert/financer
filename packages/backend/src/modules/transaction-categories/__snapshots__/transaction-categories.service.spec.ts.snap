// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionCategoriesService should return a transaction category by id 1`] = `
{
  "__v": 0,
  "_id": "623b58ada3deba9879422fbf",
  "createdAt": 2024-05-10T10:21:18.363Z,
  "deleted": false,
  "name": "Category for all types",
  "owner": "61460d7354ea082ad0256749",
  "parent_category_id": null,
  "updatedAt": 2024-05-10T10:21:18.363Z,
  "visibility": [
    "income",
    "expense",
    "transfer",
  ],
}
`;

exports[`TransactionCategoriesService should return an array of transaction categories from findAllByUser 1`] = `
[
  {
    "__v": 0,
    "_id": "623b58ada3deba9879422fbf",
    "createdAt": 2024-05-10T10:21:18.363Z,
    "deleted": false,
    "name": "Category for all types",
    "owner": "61460d7354ea082ad0256749",
    "parent_category_id": null,
    "updatedAt": 2024-05-10T10:21:18.363Z,
    "visibility": [
      "income",
      "expense",
      "transfer",
    ],
  },
  {
    "__v": 0,
    "_id": "623b58b8a3deba9879422fc2",
    "createdAt": 2024-05-10T10:21:18.364Z,
    "deleted": false,
    "name": "Invisible category",
    "owner": "61460d7354ea082ad0256749",
    "parent_category_id": null,
    "updatedAt": 2024-05-10T10:21:18.364Z,
    "visibility": [],
  },
  {
    "__v": 0,
    "_id": "623b58bfa3deba9879422fc5",
    "createdAt": 2024-05-10T10:21:18.364Z,
    "deleted": false,
    "name": "Expense category",
    "owner": "61460d7354ea082ad0256749",
    "parent_category_id": null,
    "updatedAt": 2024-05-10T10:21:18.364Z,
    "visibility": [
      "expense",
    ],
  },
  {
    "__v": 0,
    "_id": "623b58c5a3deba9879422fc8",
    "createdAt": 2024-05-10T10:21:18.364Z,
    "deleted": false,
    "name": "Income category",
    "owner": "61460d7354ea082ad0256749",
    "parent_category_id": null,
    "updatedAt": 2024-05-10T10:21:18.364Z,
    "visibility": [
      "income",
    ],
  },
  {
    "__v": 0,
    "_id": "623b6b17a3deba9879422fcd",
    "createdAt": 2024-05-10T10:21:18.364Z,
    "deleted": false,
    "name": "Transfer category",
    "owner": "61460d7354ea082ad0256749",
    "parent_category_id": null,
    "updatedAt": 2024-05-10T10:21:18.364Z,
    "visibility": [
      "transfer",
    ],
  },
  {
    "__v": 0,
    "_id": "623b6b36a3deba9879422fd1",
    "createdAt": 2024-05-10T10:21:18.364Z,
    "deleted": false,
    "name": "Transfer sub category",
    "owner": "61460d7354ea082ad0256749",
    "parent_category_id": "623b58b8a3deba9879422fc2",
    "updatedAt": 2024-05-10T10:21:18.364Z,
    "visibility": [
      "transfer",
    ],
  },
  {
    "__v": 0,
    "_id": "623b6b51a3deba9879422fd5",
    "createdAt": 2024-05-10T10:21:18.364Z,
    "deleted": false,
    "name": "Income sub category",
    "owner": "61460d7354ea082ad0256749",
    "parent_category_id": "623b58b8a3deba9879422fc2",
    "updatedAt": 2024-05-10T10:21:18.364Z,
    "visibility": [
      "income",
    ],
  },
  {
    "__v": 0,
    "_id": "623b6b64a3deba9879422fd9",
    "createdAt": 2024-05-10T10:21:18.364Z,
    "deleted": false,
    "name": "Expense sub category",
    "owner": "61460d7354ea082ad0256749",
    "parent_category_id": "623b58b8a3deba9879422fc2",
    "updatedAt": 2024-05-10T10:21:18.364Z,
    "visibility": [
      "expense",
    ],
  },
  {
    "__v": 0,
    "_id": "623b6b84a3deba9879422fdd",
    "createdAt": 2024-05-10T10:21:18.364Z,
    "deleted": false,
    "name": "Invisble sub category",
    "owner": "61460d7354ea082ad0256749",
    "parent_category_id": "623b58ada3deba9879422fbf",
    "updatedAt": 2024-05-10T10:21:18.364Z,
    "visibility": [],
  },
  {
    "__v": 0,
    "_id": "623b6b91a3deba9879422fe1",
    "createdAt": 2024-05-10T10:21:18.364Z,
    "deleted": false,
    "name": "Sub category for all types",
    "owner": "61460d7354ea082ad0256749",
    "parent_category_id": "623b58b8a3deba9879422fc2",
    "updatedAt": 2024-05-10T10:21:18.364Z,
    "visibility": [
      "income",
      "expense",
      "transfer",
    ],
  },
]
`;

exports[`TransactionCategoriesService should return an array of transaction categories from findMonthlySummariesByUserAndId 1`] = `
[
  {
    "_id": {
      "month": 1,
      "year": 2022,
    },
    "expense": {
      "amount": 1135.98,
      "count": 2,
      "transactionAmount": 2494.16,
    },
    "income": {
      "amount": 2271.5299999999997,
      "count": 2,
      "transactionAmount": 3367.1099999999997,
    },
    "total": {
      "amount": 1135.55,
      "count": 4,
      "transactionAmount": 872.9499999999998,
    },
    "transfer": {
      "amount": 0,
      "count": 0,
      "transactionAmount": 0,
    },
  },
]
`;
